Docs (GraphQL): 21.03 feature documentation for lambda webhooks (#126)
dgraph-io/dgraph-docs: content/graphql/lambda/webhook.md

+++
title = "Lambda Webhooks"
description = "Ready to use lambdas for webhooks? This documentation takes you through the schemas, resolvers, and examples."
weight = 5
[menu.main]
    parent = "lambda"
+++

### Schema

> To set up a lambda webhook, first you need to define it in your GraphQL schema by using the `@lambdaOnMutate` directive along with the mutation events (add/update/delete) you want to listen on.

Reviewer: ```suggestion
To set up a lambda webhook, you need to define it in your GraphQL schema by using the `@lambdaOnMutate` directive along with the mutation events (`add`/`update`/`delete`) you want to listen on.
```

> For example, to define a lambda webhook for all mutation events (add/update/delete) on any `Author` object:

Reviewer: ```suggestion
For example, to define a lambda webhook for all mutation events (`add`/`update`/`delete`) on any `Author` object:
```

```graphql
type Author @lambdaOnMutate(add: true, update: true, delete: true) {
    id: ID!
    name: String! @search(by: [hash, trigram])
    dob: DateTime
    reputation: Float
}
```

### Resolver

Once the schema is ready, you can define your JavaScript functions and add those as resolvers in your JS source code. 
To add the resolvers you should use the `addWebHookResolvers`method.

{{% notice "note" %}}
A Lambda Webhook resolver can use a combination of `event`, `dql`, `graphql` or `authHeader` inside the function.
{{% /notice %}}

> You have access to the `event` object within the resolver. The definition of `event` is something as follows:

Reviewer: ```suggestion
#### Event object

You also have access to the `event` object within the resolver. The definition of `event` is as follows:
```

```
"event": {
    "__typename": "<Typename>",
    "operation": "<one-of: add/update/delete>",
    "commitTs": <uint64, the commitTs of the mutation>
    "add": {
      "rootUIDs": [<list-of-UIDs-that-were-created-for-root-nodes-in-this-mutation>],
      "input": [<AddTypeInput: i.e. all the data that was received as part of the `input` argument>]
    },
    "update": {
      "rootUIDs": [<list-of-UIDs-of-root-nodes-for-which-somethig-was-set/removed-in-this-mutation>],
      "setPatch": <TypePatch: the object that was received as the patch for set>,
      "removePatch": <TypePatch: the object that was received as the patch for remove>
    },
    "delete": {
      "rootUIDs": [<list-of-UIDs-of-root-nodes-which-were-deleted-in-this-mutation>]
    }
```

> For example, to define the JavaScript lambda functions for each mutation event for which `@lambdaOnMutate` is enabled and add those as resolvers:

Reviewer: ```suggestion
#### Resolver examples

For example, to define JavaScript lambda functions for each mutation event for which `@lambdaOnMutate` is enabled and add those as resolvers:
```

```javascript
async function addAuthorWebhook({event, dql, graphql, authHeader}) {
    // execute what you want on addition of an author 
    // maybe send a welcome mail to the author
    
}

async function updateAuthorWebhook({event, dql, graphql, authHeader}) {
    // execute what you want on updation of an author
    // maybe send a mail to the author informing that few details have been updated 
    
}

async function deleteAuthorWebhook({event, dql, graphql, authHeader}) {
    // execute what you want on deletion of an author
    // maybe mail the author saying they have been removed from the platform 
    
}

self.addWebHookResolvers({
    "Author.add": addAuthorWebhook,
    "Author.update": updateAuthorWebhook,
    "Author.delete": deleteAuthorWebhook,
})
```

### Example

> Finally, if you execute a `addAuthor` mutation the resolver `addAuthorWebhook` mapped to `add` operation will get triggered :

Reviewer: ```suggestion
Finally, if you execute an `addAuthor` mutation the resolver `addAuthorWebhook` mapped to the `add` operation will get triggered:
```

+++
title = "Lambda Webhooks"
description = "Ready to use lambdas for webhooks? This documentation takes you through the schemas, resolvers, and examples."
weight = 5
[menu.main]
    parent = "lambda"
+++

### Schema

To set up a lambda webhook, you need to define it in your GraphQL schema by using the `@lambdaOnMutate` directive along with the mutation events (`add`/`update`/`delete`) you want to listen on.

For example, to define a lambda webhook for all mutation events (`add`/`update`/`delete`) on any `Author` object:

```graphql
type Author @lambdaOnMutate(add: true, update: true, delete: true) {
    id: ID!
    name: String! @search(by: [hash, trigram])
    dob: DateTime
    reputation: Float
}
```

### Resolver

Once the schema is ready, you can define your JavaScript functions and add those as resolvers in your JS source code. 
To add the resolvers you should use the `addWebHookResolvers`method.

{{% notice "note" %}}
A Lambda Webhook resolver can use a combination of `event`, `dql`, `graphql` or `authHeader` inside the function.
{{% /notice %}}

#### Event object

You also have access to the `event` object within the resolver. The definition of `event` is as follows:

> 

Reviewer 2: we can mention that based on the value of `operation`, only one of add, update or delete keys will be present inside the `event` object. i.e., there will always be `event[event["operation"]]` and nothing else.

Author: Done
